symengine/integer.cpp

```diff
@@ -10,10 +10,9 @@ Integer::Integer(int i)
 
 std::size_t Integer::__hash__() const
 {
-    std::hash<long long int> hash_fn;
     // only the least significant bits that fit into "signed long int" are
     // hashed:
-    return hash_fn(this->i.get_si());
+    return this->i.get_ui() * mpz_sgn(this->i.get_mpz_t());
```

Comment:
So why is this change still needed? You showed that the hashing is actually implementation dependent, and so we can't make our hashes equal to the Python hashes.

So it seems the motivation for this change might be speed, or perhaps it's not even faster? Either way, if it is not related to this PR, I would not do this as par of this PR.